### PR TITLE
Improve CompilationException messaging for generic test Jinja rendering

### LINKED ISSUE #5294: [CT-683] [Bug] 'dbt_utils' is undefined error despite it being in packages.yml + running clean / deps before compile
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

My team is attempting to upgrade to dbt v1.0 and have been following [this blog post](https://docs.getdbt.com/blog/upgrade-dbt-without-fear). I've upgraded to dbt locally and systematically made necessary changes to the codebase (including updating dependency versions in `packages.yml`. 

I've hit a wall in when I run `dbt compile`, I get the following error:

This happens after successfully running `dbt clean` and `dbt deps`.
```
❯ dbt clean; dbt deps
23:48:33  Running with dbt=1.0.4
23:48:33  Checking target/*
23:48:33  Cleaned target/*
23:48:33  Checking dbt_packages/*
23:48:33  Cleaned dbt_packages/*
23:48:33  Finished cleaning all paths.
23:48:36  Running with dbt=1.0.4
23:48:41  Installing calogica/dbt_expectations
23:48:43    Installed from version 0.5.6
23:48:43    Up to date!
23:48:43  Installing tnightengale/dbt_meta_testing
23:48:44    Installed from version 0.3.5
23:48:44    Up to date!
23:48:44  Installing dbt-labs/codegen
23:48:44    Installed from version 0.6.0
23:48:44    Up to date!
23:48:44  Installing calogica/dbt_date
23:48:44    Installed from version 0.5.7
23:48:44    Up to date!
23:48:44  Installing dbt-labs/dbt_utils
23:48:45    Installed from version 0.8.5
23:48:45    Up to date!
❯ dbt compile
23:49:38  Running with dbt=1.0.4
23:49:38  Partial parse save file not found. Starting full parse.
23:50:04  Encountered an error:
Compilation Error
  'dbt_utils' is undefined. This can happen when calling a macro that does not exist. Check for typos and/or install package dependencies with "dbt deps".
```

### Expected Behavior

I expect to run `dbt compile` after correctly formatting the `packages.yml` file and not see a Complilation Error about `dbt utils` being undefined.

### Steps To Reproduce

Use dbt version 1.0.4.
`packages.yml` is formatted as follows:
```
packages:
  - package: dbt-labs/dbt_utils
    version: 0.8.5
    
  - package: calogica/dbt_expectations
    version: 0.5.6

  - package: tnightengale/dbt_meta_testing
    version: 0.3.5

  - package: dbt-labs/codegen
    version: 0.6.0
```
Run `dbt clean; dbt deps; dbt compile`.
In our project, we get the error outlined in the first section. Perhaps it has to do with the partial parse piece?

### Relevant log output

```shell
============================== 2022-05-24 00:01:54.937565 | 0172c458-0caf-4f86-926c-c83bd7e9e605 ==============================
00:01:54.937565 [info ] [MainThread]: Running with dbt=1.0.4
00:01:54.937946 [debug] [MainThread]: running dbt with arguments Namespace(record_timing_info=None, debug=None, log_format=None, write_json=None, use_colors=None, printer_width=None, warn_error=None, version_check=None, partial_parse=None, single_threaded=False, use_experimental_parser=None, static_parser=None, profiles_dir='/Users/alexandrafetterman/.dbt', send_anonymous_usage_stats=None, fail_fast=None, event_buffer_size=None, project_dir=None, profile=None, target=None, vars='{}', log_cache_events=False, parse_only=False, threads=None, select=None, exclude=None, selector_name=None, state=None, full_refresh=False, defer=None, cls=<class 'dbt.task.compile.CompileTask'>, which='compile', rpc_method='compile')
00:01:54.938113 [debug] [MainThread]: Tracking: tracking
00:01:54.954641 [debug] [MainThread]: Sending event: {'category': 'dbt', 'action': 'invocation', 'label': 'start', 'context': [<snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x10cca8970>, <snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x10ccc2310>, <snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x10ccc2be0>]}
00:01:55.032942 [info ] [MainThread]: Partial parse save file not found. Starting full parse.
00:01:55.033304 [debug] [MainThread]: Sending event: {'category': 'dbt', 'action': 'partial_parser', 'label': '0172c458-0caf-4f86-926c-c83bd7e9e605', 'context': [<snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x10cca1e20>]}
```
I also get `static parser failed on modelX...` followed by `parser fallback to jinja rendering on modelX...` in the logs following, but not clear how that's related to dbt_utils.

When I run `dbt --debug --no-partial-parse compile` I get the following after a host of `static parser failed on modelX` and `parser fallback to jinja rendering on modelX` outputs.

```
00:43:16.744884 [debug] [MainThread]: Sending event: {'category': 'dbt', 'action': 'invocation', 'label': 'end', 'context': [<snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x11446ed30>, <snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x115bda640>, <snowplow_tracker.self_describing_json.SelfDescribingJson object at 0x115dafbb0>]}
00:43:16.746092 [debug] [MainThread]: Flushing usage events
00:43:17.267494 [error] [MainThread]: Encountered an error:
Compilation Error
  'dbt_utils' is undefined. This can happen when calling a macro that does not exist. Check for typos and/or install package dependencies with "dbt deps".
00:43:17.271996 [debug] [MainThread]: Traceback (most recent call last):
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/clients/jinja.py", line 516, in catch_jinja
    yield
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/clients/jinja.py", line 548, in render_template
    return template.render(ctx)
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/clients/jinja.py", line 187, in render
    return self.environment.handle_exception()
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/clients/jinja.py", line 138, in quoted_native_concat
    head = list(islice(nodes, 2))
  File "<template>", line 1, in top-level template code
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/jinja2/sandbox.py", line 407, in getattr
    value = getattr(obj, attribute)
jinja2.exceptions.UndefinedError: 'dbt_utils' is undefined

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/main.py", line 127, in main
    results, succeeded = handle_and_check(args)
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/main.py", line 192, in handle_and_check
    task, res = run_from_args(parsed)
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/main.py", line 246, in run_from_args
    results = task.run()
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/task/runnable.py", line 454, in run
    self._runtime_initialize()
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/task/runnable.py", line 156, in _runtime_initialize
    super()._runtime_initialize()
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/task/runnable.py", line 94, in _runtime_initialize
    self.load_manifest()
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/task/runnable.py", line 81, in load_manifest
    self.manifest = ManifestLoader.get_full_manifest(self.config)
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/parser/manifest.py", line 200, in get_full_manifest
    manifest = loader.load()
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/parser/manifest.py", line 345, in load
    self.parse_project(
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/parser/manifest.py", line 453, in parse_project
    parser.parse_file(block, dct=dct)
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/parser/schemas.py", line 510, in parse_file
    self.parse_tests(test_block)
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/parser/schemas.py", line 486, in parse_tests
    self.parse_column_tests(block, column)
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/parser/schemas.py", line 210, in parse_column_tests
    self.parse_test(block, test, column)
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/parser/schemas.py", line 482, in parse_test
    self.parse_node(block)
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/parser/schemas.py", line 420, in parse_node
    node = self._parse_generic_test(
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/parser/schemas.py", line 290, in _parse_generic_test
    builder = TestBuilder(
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/parser/generic_test_builders.py", line 245, in __init__
    value = get_rendered(value, render_ctx, native=True)
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/clients/jinja.py", line 597, in get_rendered
    return render_template(template, ctx, node)
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/clients/jinja.py", line 548, in render_template
    return template.render(ctx)
  File "/usr/local/Cellar/python@3.9/3.9.13/Frameworks/Python.framework/Versions/3.9/lib/python3.9/contextlib.py", line 137, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/usr/local/Cellar/dbt-redshift@1.0.0/1.0.0_2/libexec/lib/python3.9/site-packages/dbt/clients/jinja.py", line 521, in catch_jinja
    raise UndefinedMacroException(str(e), node) from e
dbt.exceptions.UndefinedMacroException: Compilation Error
  'dbt_utils' is undefined. This can happen when calling a macro that does not exist. Check for typos and/or install package dependencies with "dbt deps".
```

### Environment

- OS: MacOS Monterey Version 12.4
- Python: `python3 --version` = Python 3.9.13
- dbt: `dbt --version` 
```
installed version: 1.0.4
   latest version: 1.1.0

Your version of dbt is out of date! You can find instructions for upgrading here:
https://docs.getdbt.com/docs/installation

Plugins:
  - redshift: 1.0.0 - Update available!
  Your version of dbt-redshift is out of date! You can find instructions for upgrading here:
  https://docs.getdbt.com/dbt-cli/install/overview

  - postgres: 1.0.4 - Up to date!
  - snowflake: 0.18.2 - Out of date!
  Your version of dbt-snowflake is out of date! You can find instructions for upgrading here:
  https://docs.getdbt.com/dbt-cli/install/overview

  - redshift: 1.0.0 - Update available!
  Your version of dbt-redshift is out of date! You can find instructions for upgrading here:
  https://docs.getdbt.com/dbt-cli/install/overview

  - postgres: 1.0.4 - Up to date!
```
We're intentionally only going to v1.0 because we originally tried to jump straight to 1.1 and got these errors, so thought scaling back might resolve. No dice.

### What database are you using dbt with?

redshift

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by nicholasyager**
Resolves #5294 

### Description

In issue #5294, the reporter was having difficulties identifying the source of a `CompliationException` during a migration from 0.19.1 to 1.0.1. The original error message is as follows:

```
Compilation Error
  'dbt_utils' is undefined. This can happen when calling a macro that does not exist. Check for typos and/or install package dependencies with "dbt deps".
```

The message above does not provide any context as to where the issue was originating from.

This PR adds an exception handler for `UndefinedMacroException`s to the `TestBuilder` class so that additional context can be added to the exception raised in the event that a custom macro (or otherwise undefined macro) is reference in a generic test's configuration values. Additionally, a `CompilationException` exception handler was added to the `_parse_generic_test` method of `SchemaParser` so that the schema file that raised the `CompilationException` is also reported.

The new exception message is as follows:

```
Compilation Error
  Invalid generic test configuration given in models/example/schema.yml:
  The {target.name}.{column_name} column's "{test_name}" test references an undefined macro in its where {config_name} argument. The macro '{macro_name}' is undefined.
  Please note that the generic test configuration parser currently does not support using custom macros to populate configuration values
  	@: UnparsedNodeUpdate(original_file_path='model...ne)
```

A test has been added to confirm that a `CompilationException` is raised as expected when a custom macro is referenced in a generic test's config values, and to confirm that the appropriate exception message is returned.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#Adding-CHANGELOG-Entry)

### CODE REVIEW

File: core/dbt/parser/generic_test_builders.py [hunk 1]
```diff
@@ -257,7 +257,29 @@ def __init__(
...
+                    # and https://github.com/dbt-labs/dbt-core/issues/5294
+                    raise_compiler_error(
+                        f"\tThe {self.target.name}.{column_name} column's "
+                        f'"{self.name}" test references an undefined \n'
+                        f"\tmacro in its {key} configuration argument. "
+                        f"The macro {e.msg}.\n"
+                        "\tPlease note that the generic test configuration"
+                        " parser currently does\n"
+                        "\tusing custom macros to"
+                        " populate configuration values\n"
```
Comment (emmyoop) [inline core/dbt/parser/generic_test_builders.py:280]: Can you remove the tab formatting in this?  There should also only be a newline for the "Please note..." section.

I think you need to revisit the message.  It does not match what you laid out in the PR description.

**Review by emmyoop — CHANGES_REQUESTED** [review]
The messages need a bit more work but this is on the right track!

**Comment by nicholasyager** [comment]
Thank you for the thoughtful feedback @emmyoop! I believe I've made all of the requested changes. Please let me know if I'm missing anything! 😁

File: core/dbt/parser/generic_test_builders.py [hunk 2]
```diff
@@ -257,7 +257,28 @@ def __init__(
...
+                    # See https://github.com/dbt-labs/dbt-core/issues/4103
+                    # and https://github.com/dbt-labs/dbt-core/issues/5294
+                    raise_compiler_error(
+                        f"The {self.target.name}.{column_name} column's "
+                        f'"{self.name}" test references an undefined '
+                        f"macro in its {key} configuration argument. "
+                        f"The macro {e.msg}.\n"
+                        "Please note that the generic test configuration parser "
+                        "currently does not support using custom macros to "
+                        "populate configuration values"
```
Comment (emmyoop) [inline core/dbt/parser/generic_test_builders.py:279]: @jtcohen6 Do you have any thoughts in the wording of this exception?

Comment (dbeatty10) [reply core/dbt/parser/generic_test_builders.py:279]: This new error message format looks great to me 👍 

The thing that will be super helpful is having all the additional context to help troubleshoot 💪

**Review by emmyoop — APPROVED** [review]
Looks great!  This is a nice improvement to messaging @nicholasyager!

**Pull request merged**